objects/src/accounts/code.rs

```diff
@@ -123,36 +129,44 @@ impl AccountCode {
     /// Returns the procedure index for the procedure with the specified root or None if such
     /// procedure is not defined for this account.
     pub fn get_procedure_index_by_root(&self, root: Digest) -> Option<usize> {
-        let root_bytes = root.as_bytes();
-        self.procedures.binary_search_by(|x| x.as_bytes().cmp(&root_bytes)).ok()
+        self.procedures.iter().position(|r| r == &root)
     }
 }
 
-// SERIALIZATION
+// EQUALITY
 // ================================================================================================
 
-#[cfg(feature = "serde")]
-mod serialization {
-    use assembly::ast::{AstSerdeOptions, ModuleAst};
+impl PartialEq for AccountCode {
+    fn eq(&self, other: &Self) -> bool {
+        // TODO: consider checking equality based only on the set of procedures
+        self.module == other.module && self.procedures == other.procedures
+    }
+}
 
-    pub fn serialize<S>(module: &super::ModuleAst, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer,
-    {
-        let bytes = module.to_bytes(AstSerdeOptions {
-            serialize_imports: true,
-        });
+impl Eq for AccountCode {}
 
-        serializer.serialize_bytes(&bytes)
+// SERIALIZATION
+// ================================================================================================
+
+impl Serializable for AccountCode {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        self.module.write_into(target, MODULE_SERDE_OPTIONS);
+        self.module.write_source_locations(target);
+        // since the number of procedures is guaranteed to be between 1 and 256, we can store the
+        // number as a single byte - but we do have to subtract 1 to store 256 as 255.
+        target.write_u8(self.procedures.len() as u8 - 1);
+        self.procedures.write_into(target);
     }
+}
 
-    pub fn deserialize<'de, D>(deserializer: D) -> Result<super::ModuleAst, D::Error>
-    where
-        D: serde::Deserializer<'de>,
-    {
-        let bytes: Vec<u8> = <Vec<u8> as serde::Deserialize>::deserialize(deserializer)?;
+impl Deserializable for AccountCode {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let mut module = ModuleAst::read_from(source, MODULE_SERDE_OPTIONS)?;
+        module.load_source_locations(source)?;
+        let num_procedures = (source.read_u8()? as usize) + 1;
```

Comment:
why not `into()`? there is an infallible conversion [from `u8` to `usize`](https://doc.rust-lang.org/std/primitive.u8.html#impl-From%3Cu8%3E-for-usize)

Reply:
It seems like type inference can't figure out what is the right target type for `into()` (unless I move the `+ 1` onto a different line). So, I'm leaving it as is.